async/runtime/src/openxla/runtime/async/module.c

```diff
@@ -0,0 +1,294 @@
+}
+
+enum iree_async_runtime_module_await_pc_e {
+  IREE_ASYNC_RUNTIME_MODULE_AWAIT_PC_BEGIN = 0,
+  IREE_ASYNC_RUNTIME_MODULE_AWAIT_PC_RESUME,
+};
+
+IREE_VM_ABI_EXPORT(iree_async_runtime_module_create_async_token,  //
+                   iree_async_runtime_module_state_t,             //
+                   v, r) {
+  iree_async_value_t *token = NULL;
+  iree_status_t status = iree_async_value_create_token(&token);
+  if (iree_status_is_ok(status)) {
+    rets->r0 = iree_async_value_move_ref(token);
+  } else {
+    iree_async_value_release(token);
+  }
+
+  return status;
+}
+
+IREE_VM_ABI_EXPORT(iree_async_runtime_module_create_async_value_i32,  //
+                   iree_async_runtime_module_state_t,                 //
+                   v, r) {
+  iree_async_value_t *value = NULL;
+  iree_status_t status = iree_async_value_create_i32(&value);
+  if (iree_status_is_ok(status)) {
+    rets->r0 = iree_async_value_move_ref(value);
+  } else {
+    iree_async_value_release(value);
+  }
+
+  return status;
+}
+
+IREE_VM_ABI_EXPORT(iree_async_runtime_module_query_async_value,  //
+                   iree_async_runtime_module_state_t,            //
+                   r, i) {
+  iree_async_value_t *value = NULL;
+  IREE_RETURN_IF_ERROR(iree_async_value_check_deref(args->r0, &value));
+
+  iree_status_t query_status = iree_async_value_query(value);
+  rets->i0 = iree_status_consume_code(query_status);
+  iree_status_ignore(query_status);
+
+  return iree_ok_status();
+}
+
+IREE_VM_ABI_EXPORT(iree_async_runtime_module_signal_async_value,  //
+                   iree_async_runtime_module_state_t, r, v) {
+  iree_async_value_t *value = NULL;
+  IREE_RETURN_IF_ERROR(iree_async_value_check_deref(args->r0, &value));
+  return iree_async_value_signal(value);
+}
+
+IREE_VM_ABI_EXPORT(iree_async_runtime_module_fail_async_value,  //
+                   iree_async_runtime_module_state_t, r, v) {
+  iree_async_value_t *value = NULL;
+  IREE_RETURN_IF_ERROR(iree_async_value_check_deref(args->r0, &value));
+  iree_async_value_fail(value);
+  return iree_ok_status();
+}
+
+// Enters a wait frame for the async value |value|
+// Returns an |out_wait_status| of OK if |value| is available or
+// IREE_STATUS_DEFERRED if it is still pending and a wait
+// frame was entered.
+static iree_status_t iree_async_runtime_module_async_value_await_begin(
+    iree_vm_stack_t *stack, iree_async_value_t *value, iree_zone_id_t zone_id,
+    iree_status_t *out_wait_status) {
+  iree_vm_wait_frame_t *wait_frame = NULL;
+  IREE_RETURN_IF_ERROR(iree_vm_stack_wait_enter(
+      stack, IREE_VM_WAIT_ALL, 1, iree_infinite_timeout(), 0, &wait_frame));
+
+  iree_wait_source_t wait_source = iree_async_value_await(value);
+  wait_frame->wait_sources[0] = wait_source;
+  wait_frame->count = 1;
+
+  *out_wait_status = iree_status_from_code(IREE_STATUS_DEFERRED);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_async_runtime_module_async_value_await(
```

Comment:
I wonder can this be implemented as C++ native module?

Reply:
We need to access the stack frame in async_value_await function. I do not find ways to access stack frame in C++ native module